File: algo_9_greedy/L452-m.cpp
//
// Created by 李恒 on 2020/7/3.
//

//也是计算不重叠的区间个数，不过和 Non-overlapping Intervals 的区别在于，[1, 2] 和 [2, 3] 在本题中算是重叠区间。

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int findMinArrowShots(vector<vector<int>>& points) {
        int sz=points.size();
        if(sz == 0) return 0;
        sort(points.begin(), points.end(), cmp);
        int count=1;
        int end = points[0][1];
        for(int i=1; i<sz; ++i){
            if(points[i][0]<=end)
                continue;
            ++count;
            end = points[i][1];
        }
        return count;
    }

    static bool cmp(vector<int>&a, vector<int>&b){
        return a[1] < b[1]; //从小到大排
    }
};

File: python-algorithm/algo_10_greedy/L452-m.py
# -*- coding: utf-8 -*-

# 可以转换成435题  删除最少的区间 使剩下的区间互不重叠
# 此外，[1, 2] 和 [2, 3] 在本题中算是重叠区间

# 本题返回的是最终剩余不重叠区间的个数

class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        sz = len(points)
        if sz == 0:
            return 0
        points.sort(key = lambda x:x[1])
        ref = points[0][1]
        cnt = 0
        for i in range(1, sz):
            if points[i][0] <= ref:
                cnt += 1
            else:
                ref = points[i][1]
        return sz-cnt
